cfep-13.md

```diff
@@ -0,0 +1,64 @@
+
+<table>
+<tr><td> Title </td><td> Securing `conda-forge` Uploads to `anaconda.org` </td>
+<tr><td> Status </td><td> Proposed </td></tr>
+<tr><td> Author(s) </td><td> Matt, Marius, Anthony, Isuru</td></tr>
+<tr><td> Created </td><td> Feb 27, 2020</td></tr>
+<tr><td> Updated </td><td> Feb 27, 2020</td></tr>
+<tr><td> Discussion </td><td> NA </td></tr>
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Abstract
+
+Currently the `anaconda.org` tokens on each feedstock allow any `conda-forge`
+feedstock maintainer to upload any output to any `anaconda.org` package.
+This could allow someone to push a compromised and/or malicious build
+of say the `python` package. This situation is a security risk that we
+should mitigate.
+
+## Specification
+
+We will secure the uploads of outputs to `anaconda.org` via the following process.
+
+1. Each feedstock is provisioned with a secret key that is only known to the
+   maintainers of that feedstock and the core team. These keys will be kept in
+   a private repo with only core members having write access. This key will be pushed to
+   the CI providers for that repo as well.
+2. Feedstock creation will move to a separate process in order to secure the keys
+   associated with our CI provider accounts and to securely register outputs of feedstocks.
+3. A public, global registry of all outputs from all feedstocks will be kept
+   in a repo on github. Write access to this repo is limited to core as well.
+4. When a feedstock builds new outputs it wants to upload, it will push these outputs
+   to a staging organization on `anaconda.org`. It will also send the names of the
+   outputs, their SHA256 checksums, the feedstock name, and the secret key to a
+   validation web service. This web service then verifies the secret key-feedstock
+   combination and verifies that the outputs are allowed for that feedstock.
+4. If everything is ok, the outputs are copied from the staging organization to the
+   main conda-forge channels.
+6. The service should post a comment on the PR that the outputs were copied.
+
+
+### The Impact on Users
+
+In order to mitigate the impact on feedstock maintainers, we will
+
+1. Advertise this change widely as it is implemented.
+2. Create an admin bot command to create a PR for new outputs.
+3. Add a linter check that all outputs are registered for this feedstock in the global database.
```

Comment:
This would make the linter as slow at the rerender command because the package has to be rerendered for all the platforms.

Reply:
Hmmmm. We could allow the CI on PRs to do a test check of the outputs and the CI can fail for PRs if outputs are not registered.

Reply:
My main thing is that something somewhere needs to report in a visible way to users what they need to do.

Reply:
We could have the rerender command do this check and post a comment. That is slow anyways and actually renders the recipe.

Reply:
+1. I would remove the linter check then add something about using CI for this.

Reply:
> We could have the rerender command do this check and post a comment. That is slow anyways and actually renders the recipe.

You mean run rerender on every push to a PR? That would cripple the webservice

Reply:
No I mean when it happens to be called. So if a user happens to use the rerender service to rerender a feedstock, we go ahead and check the outputs then.

Reply:
The CI should check too.

Reply:
That's too much complexity. Just CI is enough IMO.

Reply:
fair enough - I guess it can post a loud error message that people can look at